lib/code-caller-docker.js

```diff
@@ -0,0 +1,685 @@
+                    }
+                }
+
+                await this._ensureImage();
+                const binds = [
+                    `${elementsHostDir}:/elements:ro`,
+                    `${pythonHostDir}:/python:ro`,
+                    `${courseDir}:/course:ro`,
+                ];
+                await this._createAndAttachContainer(binds);
+                this.state = WAITING;
+            } else if (config.workersExecutionMode === 'native') {
+                await this._ensureDirectoryIfNeeded();
+                await createBindMount(course.path, this.hostDirectory.path);
+            } else {
+                throw new Error(`Unexpected worker execution mode "${config.workersExecutionMode}"`);
+            }
+
+        })(callback);
+    }
+
+    /**
+     * 
+     * @param {'restart' | 'question' | 'course-element' | 'core-element'} type 
+     * @param {string} directory 
+     * @param {string} file 
+     * @param {string} fcn 
+     * @param {string[]} args 
+     * @param {(err: Error | null | undefined, data?: any, output?: string) => void} callback 
+     */
+    call(type, directory, file, fcn, args, callback) {
+        debug(`enter call(), state: ${String(this.state)}, uuid: ${this.uuid}`);
+        this.ensureChild().then(() => {
+            if (!this._checkState([WAITING])) return callback(new Error('invalid DockerCaller state'));
+            if (!this._checkReadyForCall()) return callback(new Error('not ready for call'));
+            // TODO: state checking might be off. for the 'course' allocation method, we only start
+            // containers when we know about the course. however, for the general case in prod, we can
+            // start containers immediately on boot by "warming up". Need to make sure we actually do that,
+            // and that it's handled correctly.
+
+            let cwd = null;
+            let paths = ['/python'];
+            if (type === 'question') {
+                cwd = path.posix.join('/course', 'questions', directory);
+                paths.push('/course/serverFilesCourse');
+            } else if (type === 'course-element') {
+                cwd = path.join('/course', 'elements', directory);
+                paths.push('/course/serverFilesCourse');
+            } else if (type === 'core-element') {
+                cwd = path.posix.join('/elements', directory);
+            } else if (type === 'restart') {
+                // Doesn't need a working directory
+            } else {
+                callback(new Error(`Unknown function call type: ${type}`));
+                return;
+            }
+
+            const callData = { file, fcn, args, cwd, paths };
+            const callDataString = JSON.stringify(callData);
+
+            this.output = '';
+            this.lastCallData = callData;
+            this.callback = callback;
+            // TODO: how long should timeout actually be?
```

Comment:
Ah, I just read this after making my other comment about timeouts. I’d suggest that timeout-outside-container should be equal to timeout-inside-container plus 5 seconds, or something like that. Using a fixed additional time seems more robust than a scaling (1.5x)?

Reply:
@mwest1066 I didn't change this yet - we're still at a 1.5x multiplier. Let me know if you still have thoughts about this, it's been 2 years now!

Reply:
Is there a reason why a container with a longer "internal" timeout would need a *proportionally* longer "external" timeout, or did it just coincidentally work out to good values so far?

Reply:
Yep, I think it's just coincidentally been good values so far 😅 y'all are probably right that a fixed value would be better. Will change!

Reply:
Done in b395fb0.